apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +239,707 @@ terminate(_Reason, _St) ->
-              #st{chan_id = ChanId}, C) ->
+check_req(#{mode := unlock} = R, #st{chan_id = ChId, chan_vsn = Vsn} = St, C) ->
+    if Vsn =/= undefined ->
+            lager:debug("checking", []),
+            case get_basic_ch_status_(ChId, C) of
+                {#{locked_until := 0} = Ch, C1} ->
+                    lager:debug("locked_until = 0", []),
+                    report_channel_unlocked(R, Ch, St, C1),
+                    {done, C1};
+                {#{locked_until := LockedUntil} = Ch, C1} ->
+                    {Height, C2} = top_height(C1),
+                    lager:debug("LockedUntil = ~p, Height = ~p", [LockedUntil, Height]),
+                    case LockedUntil < Height of
+                        true  ->
+                            lager:debug("locked_until expired", []),
+                            report_channel_unlocked(R, Ch, St, C2),
+                            {done, C2};
+                        false ->
+                            lager:debug("still locked", []),
+                            {R#{check_at_height => LockedUntil + 1}, C2}
+                    end;
+                {undefined, C1} ->
+                    lager:debug("couldn't find channel", []),
+                    {done, C1}
+            end;
+       true ->
+            lager:debug("Vsn = undefined", []),
+            {R, C}
+    end;
+check_req(#{mode := watch} = R, #st{chan_id = ChId} = St,
+          #{scenario := Scenario, top_hash := Hash } = C) ->
+    case Scenario of
+        {has_tx, {_, #{ type    := TxType
+                      , tx_hash := TxHash }}} ->
+            {Ch, C1} = get_channel(ChId, Hash, C),
+            {#{ height := H }, C2} = top_info(Hash, C1),
+            case Ch of
+                undefined when TxType == channel_close_mutual_tx;
+                               TxType == channel_settle_tx ->
+                    report_closed_on_chain(#{ tx_type => TxType
+                                            , tx_hash => TxHash
+                                            , block_hash => Hash
+                                            , height  => H }, R, St, C2);
+                _ ->
+                    {Status, C3} = channel_status(ChId, C2),
+                    lager:debug("Status = ~p", [Status]),
+                    watch_for_change_in_ch_status(Status, H, R, St, C3)
+            end;
+        fork_switch ->
+            watch_for_channel_change(R, St, C);
+        _ ->
+            {R, C}
+    end;
+check_req(#{mode := tx_hash, tx_hash := TxHash, min_depth := MinDepth} = R,
+              #st{chan_id = ChanId} = St, C) ->
+    #{info := #{parent := Parent}} = R,
     lager:debug("check_status(tx_hash = ~p, parent = ~p)", [TxHash, Parent]),
-    case min_depth_achieved(TxHash, MinDepth, C) of
-        {true, C1} ->
+    case current_depth(TxHash, C) of
+        {undefined, C1} ->
+            {R, C1};
+        {Depth, C1} when Depth >= MinDepth ->
             lager:debug("min_depth achieved", []),
```

Comment:
It feels the computation of the current depth does not enforce the expectation that the block hash in which the tx hash was found is an ancestor of the top block hash. So in case of fork the depth computation may be misleading (shorter fork may have more work on it). Though seems good enough. Ok.

Reply:
Hmm, when there is a `top_changed` after a fork switch, if `TxHash` was in a block that's no longer on the chain, it will be thrown back into the mempool, and `current_depth/2` will return `{undefined, C1}`. Also, the function does check whether its enclosing block is in the main chain.

Reply:
> Hmm, when there is a `top_changed` after a fork switch, if `TxHash` was in a block that's no longer on the chain, it will be thrown back into the mempool, and `current_depth/2` will return `{undefined, C1}`.

The tx may be included in the new fork too.

> Also, the function does check whether its enclosing block is in the main chain.

Correct! I had missed that. (I had probably confused block hash for tx hash.)

Good then.